### ISSUE TITLE
Install/download only if not exists

### ISSUE THREAD

**Issue by TheCelavi** [phar-io/phive#227]
We are using phive for our CI environment, one step in the build task is to download phars. 

Phars are always downloaded, even if they exists in requested versions. In those cases, download can be skipped.

**Comment by theseer** [comment]
This is actually how phive was designed: All phars are stored in `~/.phive/phars`.  If they are being re-downloaded, that means phive can't find the files previously downloaded.

That's actually a common problem with CI systems that spawn new instances whithout keeping/sharing phive's home (`~/.phive`).

Not sure there is much we can do here without futher details, though..

**Comment by TheCelavi** [comment]
Project which is built on Jenkins is: https://github.com/RunOpenCode/doctrine-naming-strategy-bundle

Gradle script is not the one we use, I have to remove it, but it is similar, we just run an ant script to start a build. 

Phive configuration is here: https://github.com/RunOpenCode/doctrine-naming-strategy-bundle/blob/master/phive.xml, in ant script, we run "phive install" only. 

Jenkins preserves installed phars, phars are within ./tools directory of the project (as far as I know), even though it is within docker container, project directory is mounted into container. 

And we do get this output in logs, phive does download. (BTW, your tool is awesome! Thank you!)

phive-install:
     [exec] Phive 0.13.2 - Copyright (C) 2015-2020 by Arne Blankerts, Sebastian Heuer and Contributors
     [exec] Fetching repository list
     [exec] Downloading https://phar.io/data/repositories.xml
     [exec] 
     [exec] [2K ╰|>                                       | 0.00 KB / 6.04 KB -   0%[2K ╰|========================================>| 6.04 KB / 6.04 KB - 100%[2KDownloading https://phar.phpunit.de/phive.xml
     [exec] [2K ╰|====>                                   | 15.57 KB / 160.43 KB -  10%[2K ╰|========>                               | 31.55 KB / 160.43 KB -  20%[2K ╰|============>                           | 47.98 KB / 160.43 KB -  30%[2K ╰|================>                       | 64.00 KB / 160.43 KB -  40%[2K ╰|====================>                   | 79.98 KB / 160.43 KB -  50%[2K ╰|========================>               | 95.96 KB / 160.43 KB -  60%[2K ╰|============================>           | 111.95 KB / 160.43 KB -  70%[2K ╰|================================>       | 128.00 KB / 160.43 KB -  80%[2K ╰|====================================>   | 143.98 KB / 160.43 KB -  90%[2K ╰|========================================>| 160.43 KB / 160.43 KB - 100%[2KDownloading https://phar.phpunit.de/phpunit-8.5.2.phar
     [exec] [2K ╰|>                                       | 0.02 MB / 2.78 MB -   1%[2K ╰|>                                       | 0.05 MB / 2.78 MB -   2%[2K ╰|=>                                      | 0.08 MB / 2.78 MB -   3%[2K ╰|=>                                      | 0.11 MB / 2.78 MB -   4%[2K ╰|==>                                     | 0.14 MB / 2.78 MB -   5%[2K ╰|==>                                     | 0.16 MB / 2.78 MB -   6%[2K ╰|==>                                     | 0.19 MB / 2.78 MB -   7%[2K ╰|===>                                    | 0.22 MB / 2.78 MB -   8%[2K ╰|===>                                    | 0.25 MB / 2.78 MB -   9%[2K ╰|====>                                   | 0.27 MB / 2.78 MB -  10%[2K ╰|====>                                   | 0.30 MB / 2.78 MB -  11%[2K ╰|====>                                   | 0.33 MB / 2.78 MB -  12%[2K ╰|=====>                                  | 0.36 MB / 2.78 MB -  13%[2K ╰|=====>                                  | 0.39 MB / 2.78 MB -  14%[2K ╰|======>                                 | 0.41 MB / 2.78 MB -  15%[2K ╰|======>                                 | 0.44 MB / 2.78 MB -  16%[2K ╰|======>                                 | 0.47 MB / 2.78 MB -  17%[2K ╰|=======>                                | 0.50 MB / 2.78 MB -  18%[2K ╰|=======>                                | 0.52 MB / 2.78 MB -  19%[2K ╰|========>                               | 0.55 MB / 2.78 MB -  20%[2K ╰|========>                               | 0.58 MB / 2.78 MB -  21%[2K ╰|========>                               | 0.61 MB / 2.78 MB -  22%[2K ╰|=========>                              | 0.64 MB / 2.78 MB -  23%[2K ╰|=========>                              | 0.66 MB / 2.78 MB -  24%[2K ╰|==========>                             | 0.69 MB / 2.78 MB -  25%[2K ╰|==========>                             | 0.72 MB / 2.78 MB -  26%[2K ╰|==========>                             | 0.75 MB / 2.78 MB -  27%[2K ╰|===========>                            | 0.77 MB / 2.78 MB -  28%[2K ╰|===========>                            | 0.80 MB / 2.78 MB -  29%[2K ╰|============>                           | 0.83 MB / 2.78 MB -  30%[2K ╰|============>                           | 0.86 MB / 2.78 MB -  31%[2K ╰|============>                           | 0.89 MB / 2.78 MB -  32%[2K ╰|=============>                          | 0.91 MB / 2.78 MB -  33%[2K ╰|=============>                          | 0.94 MB / 2.78 MB -  34%[2K ╰|==============>                         | 0.97 MB / 2.78 MB -  35%[2K ╰|==============>                         | 1.00 MB / 2.78 MB -  36%[2K ╰|==============>                         | 1.02 MB / 2.78 MB -  37%[2K ╰|===============>                        | 1.05 MB / 2.78 MB -  38%[2K ╰|===============>                        | 1.08 MB / 2.78 MB -  39%[2K ╰|================>                       | 1.11 MB / 2.78 MB -  40%[2K ╰|================>                       | 1.14 MB / 2.78 MB -  41%[2K ╰|================>                       | 1.16 MB / 2.78 MB -  42%[2K ╰|=================>                      | 1.19 MB / 2.78 MB -  43%[2K ╰|=================>                      | 1.22 MB / 2.78 MB -  44%[2K ╰|==================>                     | 1.25 MB / 2.78 MB -  45%[2K ╰|==================>                     | 1.27 MB / 2.78 MB -  46%[2K ╰|==================>                     | 1.30 MB / 2.78 MB -  47%[2K ╰|===================>                    | 1.33 MB / 2.78 MB -  48%[2K ╰|===================>                    | 1.36 MB / 2.78 MB -  49%[2K ╰|====================>                   | 1.39 MB / 2.78 MB -  50%[2K ╰|====================>                   | 1.41 MB / 2.78 MB -  51%[2K ╰|====================>                   | 1.44 MB / 2.78 MB -  52%[2K ╰|=====================>                  | 1.47 MB / 2.78 MB -  53%[2K ╰|=====================>                  | 1.50 MB / 2.78 MB -  54%[2K ╰|======================>                 | 1.52 MB / 2.78 MB -  55%[2K ╰|======================>                 | 1.55 MB / 2.78 MB -  56%[2K ╰|======================>                 | 1.58 MB / 2.78 MB -  57%[2K ╰|=======================>                | 1.61 MB / 2.78 MB -  58%[2K ╰|=======================>                | 1.64 MB / 2.78 MB -  59%[2K ╰|========================>               | 1.66 MB / 2.78 MB -  60%[2K ╰|========================>               | 1.69 MB / 2.78 MB -  61%[2K ╰|========================>               | 1.72 MB / 2.78 MB -  62%[2K ╰|=========================>

**Comment by theseer** [comment]
The `tools` directory is considered a working copy for phive and not a source to get phars from in case `phive install` (or `phive update`) is requested. So preserving it, doesn't help and is not even needed as any actual content is ignored.

Phive *will* reuse existing phars from its storage directory, which is `~/.phive/phars`. If any directory should be preserved, it's `~/.phive` (or whatever alterantive directory you may want to specify as phive's home directory) as it contains the phars as well as required meta data. In that regard it's pretty much like the composer cache directory.

**Comment by TheCelavi** [comment]
I thought that in XML, specifying "location="./tools/phpunit.phar" I have explicitly said where I want to store downloaded phar. 

Is there anyway to specify similar functionality - for CI, it is kinda important to have isolated, a local copy of necessary phars (like composer does with vendor directory).

**Comment by theseer** [comment]
Hold on, you're various confusing things here I think. Let me try again ;)

The `location` attribute in the `phive.xml` is the location where phive will *install* the respective phar *into* (= destination). But to do that, it has to get the phar from somewhere (= source). The first place to look for a phar is phive's "private" storage home directory (one might call that a cache). If the file is not there, it will try to download it from remote.

That's almost exactly the same thing for composer: the `vendor` directory is working copy. If you want to install something, composer will check if the required version of that something is in `.cache/composer` and, if found, will install (= copy to vendor) from there. 

The only thing that might be different compared to composer is that we simply ignore the content of the destination and merely recopy (and potentially overwrite) existing files whereas composer ensures to not do that in case you have locally modified files. Since our default mode of operation is to create symbolic links, there is no such thing for phive. Only if you manually copy a phar into the `tools` directory, there's a chance of conflict. We considered that too small a problem to justify writing code for that at this point.

So, again: Ensure that `~/.phive` is preserved (= shared) over builds and phive will not download but merely (re-)copy the phar files into the `tools` directory.

**Comment by TheCelavi** [comment]
> The only thing that might be different compared to composer is that we simply ignore the content of the destination and merely recopy (and potentially overwrite) existing files whereas composer ensures to not do that in case you have locally modified files.

Yep, that was a confusion. I thought that this library has same method of operating as composer.

You might expect more and more people using your library as I do. Until then, I am closing this issue. Thank you, you were very helpful!

Did I mentioned that your tool is awesome?

**Comment by theseer** [comment]
Thanks for the praise!

We're aware of the fact phive's sometimes pragmatic or even out right naive approach on things might not be the most optimal for all cases. That we ignore copied, existing phars probably is among those.

Unfortunately, neither Sebastian nor I do have much time to work on our open source projects currently and the corona virus stuff doesn't help either. So some things are slow in progressing and the aforementioned approach of not retaining `tools` but rather `~/.phive` seems a good enough work around for the time being.

Btw: If you don't like it to be `~/.phive`, you can specify a custom home direcotry (phive --home /your/path/of/choice .... ).

**Comment by TheCelavi** [comment]
**Just info for anyone who stumbles upon this issue**

By using CI within docker/docker-compose, solution is:
1. Mount `${HOME}/.phive` directory into container by using same destination, ie: `${HOME}/.phive:${HOME}/.phive`
2. Run script with following flag: `phive --home ${HOME}/.phive install`

If you have multiple projects using phive to manage phars, all will benefit from same cache.
